pkg/devfile/adapters/docker/component/adapter.go

```diff
@@ -0,0 +1,219 @@
+	}
+
+	for _, comp := range supportedComponents {
+		// Check to see if this component is already running and if so, update it
+		// If component isn't running, re-create it, as it either may be new, or crashed.
+		containers, err := a.Client.GetContainersByComponentAndAlias(componentName, *comp.Alias)
+		if err != nil {
+			return errors.Wrapf(err, "unable to list containers for component %s", componentName)
+		}
+		if len(containers) == 0 {
+			// Container doesn't exist, so need to pull its image (to be safe) and start a new container
+			s := log.Spinner("Pulling image " + *comp.Image)
+			defer s.End(false)
+
+			err = a.Client.PullImage(*comp.Image)
+			if err != nil {
+				return errors.Wrapf(err, "Unable to pull %s image", *comp.Image)
+			}
+			s.End(true)
+
+			// Start the container
+			err = a.startContainer(componentName, projectVolumeName, comp)
+			if err != nil {
+				return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+			}
+			glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+		} else {
+			// Container already exists
+			containerID := containers[0].ID
+
+			// Get the associated container config from the container ID
+			containerConfig, err := a.Client.GetContainerConfig(containerID)
+			if err != nil {
+				return errors.Wrapf(err, "unable to get the container config for component %s", componentName)
+			}
+
+			// See if the container needs to be updated
+			if utils.DoesContainerNeedUpdating(comp, containerConfig) {
+				glog.V(3).Info("Updating the component")
+				// Remove the container
+				err := a.Client.RemoveContainer(containerID)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to remove container %s for component %s", containerID, *comp.Alias)
+				}
+
+				// Start the container
+				err = a.startContainer(componentName, projectVolumeName, comp)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+				}
+				glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+			}
+		}
+	}
+	return nil
+}
+
+func (a Adapter) startContainer(componentName string, projectVolumeName string, comp versionsCommon.DevfileComponent) error {
+	containerConfig := a.getContainerConfig(componentName, comp)
+	hostConfig := container.HostConfig{}
+
+	// If the component set `mountSources` to true, add the source volume to it
+	if comp.MountSources {
+		addProjectVolumeToComp(projectVolumeName, &hostConfig)
+	}
+
+	// Create the docker container
+	s := log.Spinner("Starting container for " + *comp.Image)
+	defer s.End(false)
+	err := a.Client.StartContainer(&containerConfig, &hostConfig, nil)
+	if err != nil {
+		return err
+	}
+	s.End(true)
+	return nil
+}
+func (a Adapter) getContainerConfig(componentName string, comp versionsCommon.DevfileComponent) container.Config {
```

Comment:
Nit: can we rename this function to something else, I was confused for a moment between this and lclient.GetContainerConfig which gets it from an active container and this function generates and returns the config..

Reply:
Haha, fair enough. How about `getConfigFromContainer`?

Reply:
`generateAndGetContainerConfig`? since this function generates and gets the container config

Reply:
Works for me.

Reply:
Done